### PR TITLE
Feat/remove mitt

### LINKED ISSUE #419: Proposal: giving up a "mitt"
I think "mitt" could be given up. At the moment it's used exclusively to change parent stage through events. This task can be accomplished by reactive provide/inject functionality which is well-described in documentation: https://v3.vuejs.org/guide/composition-api-provide-inject.html#mutating-reactive-properties

In our case take a form example. Right now Form provides mitt instance to make FormItem to be able to call 2 events on it: addField and removeField.

Instead of this we could provide 2 functions from Form (addField and removeField) which are to be called in FormItem to mutate parent state.

In that case we would give up needless dependency and would use vue functionality.

It would be looks like this.
In the form:

image
In the form-item
image
image

If you agree with the idea, I could make a PR a little bit later.

### PR DESCRIPTION
**Pull request by purepear**
# Remove `mitt` as build dependency by using provide/inject instead.

Discussion https://github.com/element-plus/element-plus/discussions/3158

This PR is slightly based on https://github.com/element-plus/element-plus/pull/431

_It'll probably be easier for reviewers do the initial review commit by commit_

## Affected components: 
* collapse / collapse-item
* menu / submenu / menu-item 
* tree (drag n drop)
* select / option-group / option: we are using `shallowRef` and `triggerRef` to simulate the existing mitt behavior but i imagine the **select filtering** functionality could be refactored in the future(and probably not rely on shallowRef/triggerRef as they seem slightly difficult from maintainability point of view)
* form / form-item: refactor the `validate()` functionality requires small changes in most form controls(cascader, checkbox, color-picker, input, radio, select, slider, switch, time-picker, transfer)

I removed `mitt` from 
* `packages/components/package.json`
* `packages/element-plus/package.json`
* `packages/tokens/package.json`

I left `mitt` in the root `package.json` as it is used by the `website` events [bus](https://github.com/element-plus/element-plus/blob/dev/website/bus.js) in order to fade in/out the website navigation. 

I believe removing mitt from the packages should be enough to remove it from the library build.

I also saw mitt in the [webpack build config](https://github.com/element-plus/element-plus/blob/dev/build/webpack.config.js#L30) 🤔 ~Shall i remove  it?~ **I removed it.**

- [x] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by JeremyWuuuuu** [comment]
@sxzz Can you follow through this PR review process? I will check this PR once you have done the reviewing.

**Comment by sxzz** [comment]
Of course.

File: packages/components/collapse/src/collapse-item.vue [hunk 1]
```diff
@@ -49,7 +49,7 @@
 </template>
 <script lang="ts">
 import { defineComponent, PropType, inject, computed, ref } from 'vue'
-import { CollapseProvider } from './collapse'
+import type { CollapseProvider } from './collapse.type'
```
Comment (sxzz) [inline packages/components/collapse/src/collapse-item.vue:52]: Could you please move this line 52 to 55?

**Review by sxzz — APPROVED** [review]

File: packages/components/collapse/src/collapse-item.vue (hunk 1 above)
Comment (purepear) [reply packages/components/collapse/src/collapse-item.vue]: Yep, makes sense to move `import type` after the imports 👍 . Done :)

**Comment by sxzz** [comment]
@JeremyWuuuuu Review please.

File: packages/components/form/src/form-item.vue [hunk 2]
```diff
@@ -296,43 +289,20 @@ export default defineComponent({
...
-
     const elFormItem = reactive({
       ...toRefs(props),
       size: sizeClass,
       validateState,
       $el: formItemRef,
-      formItemMitt,
-      removeValidateEvents,
-      addValidateEvents,
+      evaluateValidationEnabled,
```
Comment (JeremyWuuuuu) [inline packages/components/form/src/form-item.vue:305]: Praise: This is good!

File: packages/components/form/src/form-item.vue [hunk 3]
```diff
@@ -258,16 +256,11 @@ export default defineComponent({
...
       } else {
         prop.o[prop.k] = initialValue
       }
-      // reset validateDisabled after onFieldChange triggered
-      nextTick(() => {
-        validateDisabled.value = false
-      })
     }
 
     const getRules = () => {
```
Comment (JeremyWuuuuu) [inline packages/components/form/src/form-item.vue:266]: I was just wondering can we make this one as `computed` since we could be invoking this for a lot of times.

NB: we can and we should put this into separate PRs

File: packages/components/select/src/useOption.ts [hunk 4]
```diff
@@ -3,10 +3,12 @@ import {
   computed,
   getCurrentInstance,
   watch,
-  onBeforeUnmount,
+  Ref,
```
Comment (JeremyWuuuuu) [inline packages/components/select/src/useOption.ts:6]: **Suggestion**:  we can move this to 

```ts
import type { Ref } from 'vue'
```

File: packages/components/select/src/useOption.ts [hunk 5]
```diff
@@ -3,10 +3,12 @@ import {
...
+  toRaw,
+  unref,
 } from 'vue'
 import { getValueByPath, escapeRegexpString } from '@element-plus/utils/util'
-import { selectKey, selectGroupKey, selectEvents } from './token'
+import { selectKey, selectGroupKey, QueryChangeCtx } from './token'
```
Comment (JeremyWuuuuu) [inline packages/components/select/src/useOption.ts:11]: Same as above

File: packages/components/select/src/useSelect.ts [hunk 6]
```diff
@@ -253,8 +262,10 @@ export const useSelect = (props, states: States, ctx) => {
           }
           handleQueryChange(states.query)
           if (!props.multiple && !props.remote) {
-            states.selectEmitter.emit('elOptionQueryChange', '')
-            states.selectEmitter.emit('elOptionGroupQueryChange')
+            queryChange.value.query = ''
+
+            triggerRef(queryChange)
+            triggerRef(groupQueryChange)
```
Comment (JeremyWuuuuu) [inline packages/components/select/src/useSelect.ts:268]: **Praise:** I love this idea! 👍

**Review by jw-foss — APPROVED** [review]

**Comment by JeremyWuuuuu** [comment]
The code looks legit to me. I did leave a few comments, you can decide on your own if you want to change it. I think the change in `webpack.config.js` can be reverted since we do not use that for bundling. :bow:

This is really meaningful to us, thank you for your contribution. 👍

File: packages/components/form/src/form-item.vue (hunk 3 above)
Comment (purepear) [reply packages/components/form/src/form-item.vue:266]: Yes, i agree, sth we can do in another PR :)

**Comment by JeremyWuuuuu** [comment]
Let's make the roll!
